src/Sulu/Component/CustomUrl/Manager/CustomUrlManager.php

```diff
@@ -0,0 +1,158 @@
+{
+    /**
+     * @var DocumentManagerInterface
+     */
+    private $documentManager;
+
+    /**
+     * @var CustomUrlRepository
+     */
+    private $customUrlRepository;
+
+    /**
+     * @var PathBuilder
+     */
+    private $pathBuilder;
+
+    public function __construct(
+        DocumentManagerInterface $documentManager,
+        CustomUrlRepository $customUrlRepository,
+        PathBuilder $pathBuilder
+    ) {
+        $this->documentManager = $documentManager;
+        $this->customUrlRepository = $customUrlRepository;
+        $this->pathBuilder = $pathBuilder;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function create($webspaceKey, array $data, $locale = null)
+    {
+        $document = $this->documentManager->create('custom_urls');
+        $this->bind($document, $data, $locale);
+
+        $this->documentManager->persist(
+            $document,
+            $locale,
+            [
+                'parent_path' => $this->getItemsPath($webspaceKey),
+                'node_name' => Urlizer::urlize($document->getTitle()),
+                'load_ghost_content' => true,
+            ]
+        );
+
+        return $document;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function readList($webspaceKey)
+    {
+        // TODO pagination
+
+        return $this->customUrlRepository->findList($this->getItemsPath($webspaceKey));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function read($uuid, $locale = null)
+    {
+        return $this->documentManager->find($uuid, $locale, ['load_ghost_content' => true]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function update($uuid, array $data, $locale = null)
+    {
+        $document = $this->read($uuid, $locale);
+        $this->bind($document, $data, $locale);
+
+        $this->documentManager->persist($document, $locale);
+
+        return $document;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function delete($uuid)
+    {
+        $this->documentManager->remove($this->read($uuid));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFields()
+    {
+        return [
+            'title' => ['property' => 'title'],
+            'published' => ['property' => 'published'],
+            'baseDomain' => ['property' => 'baseDomain'],
+            'domainParts' => ['property' => 'domainParts', 'type' => 'json_array'],
+            'target' => ['property' => 'target', 'type' => 'reference'],
+            'multilingual' => ['property' => 'multilingual'],
+            'canonical' => ['property' => 'canonical'],
+            'redirect' => ['property' => 'redirect'],
+            'targetLocale' => ['property' => 'targetLocale'],
+        ];
+    }
+
+    /**
+     * Bind data array to given document.
+     *
+     * @param CustomUrlDocument $document
+     * @param array $data
+     * @param string $locale
+     */
+    private function bind(CustomUrlDocument $document, $data, $locale)
```

Comment:
Disagree that it should be handled by the DocumentManager.

The point is the request data needs to be converted to values which are
accepted by the domain object, this responsiblity is absolutely distinct
from that of managing documents. DocumentManager should never know
about the request data.

<b>Nothing "beyond" the controller should know about the request data.</b>

The Form component maps request data to domain objects - this is core
functionality of the form component (it was not extended to do this). We
would be reproducing half of this code in an own implementation. Saying
that the point is not about the form component, lets forget about the
Form component.

The data has to be normalized and validated before entering our system.
Binding here does not validate the incoming data, and nor
would it be a good idea to validate it here -- we would add boilerplate
code all over the place. 

Request mapping and validation does not even need to be coded: we can
map the request data automatically based on the Document mapping (thus
implicitly declaring a standard format expected from the UI JS code).
This is currently the case when you save a node in the "new" content
mapper. A form is automatically generated for the structure type, you
don't have to do anything.

For the Manager argument -- if we remove the Mapping and Binding from
it, it becomes essentially a Repository (albeit a read/write one), as we
already have a repository, then this class is not required.

We should consider that the document in question (i.e. the
CustomUrlDocument) could be updated from multiple places - not only the
request. Should this manager class be used then? The answer should be
"No". Whatever way you look at it, updating a document by passing a raw
array is a bad move.

On Fri, Jan 22, 2016 at 12:40:28AM -0800, Wachter Johannes wrote:
>    In [1]src/Sulu/Component/CustomUrl/Manager/CustomUrlManager.php:
> 
>  > +            'target' => ['property' => 'target', 'type' => 'reference'],
>  > +            'multilingual' => ['property' => 'multilingual'],
>  > +            'canonical' => ['property' => 'canonical'],
>  > +            'redirect' => ['property' => 'redirect'],
>  > +            'targetLocale' => ['property' => 'targetLocale'],
>  > +        ];
>  > +    }
>  > +
>  > +    /**
>  > +     * Bind data array to given document.
>  > +     *
>  > +     * @param CustomUrlDocument $document
>  > +     * @param array $data
>  > +     * @param string $locale
>  > +     */
>  > +    private function bind(CustomUrlDocument $document, $data, $locale)
> 
>    but for my opinion the controller should not know about the
>    document-manager at all and i think also not for the mapping component but
>    thats a different discussion. a manager should know about dependencies for
>    the changing entity (see hexagonal architecture talk from noback).
> 
>    For me it feels wrong to use the form component here because it has a
>    different approach which says the name form to me. It was implement to
>    handle forms and then it was extended to map data. we should use a
>    component which only maps data and i think this should be handled by the
>    document manager (then its not important which component it does - because
>    its abstracted).
> 
>    For example we could introduce $documentManager->bind($alias, $data,
>    [$identifier]) which binds the data to a new or existing document with
>    alias and optional the $identifier.
> 
>    —
>    Reply to this email directly or [2]view it on GitHub.
> 
>    Reverse link: [3]unknown
> 
> References
> 
>    Visible links
>    1. https://github.com/sulu-io/sulu/pull/1927#discussion_r50512404
>    2. https://github.com/sulu-io/sulu/pull/1927/files#r50512404
>    3. https://github.com/sulu-io/sulu/pull/1927/files#r50512404